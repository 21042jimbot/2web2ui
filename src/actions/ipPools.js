import sparkpostApiRequest from './helpers/sparkpostApiRequest';

export function listPools() {
  return sparkpostApiRequest({
    type: 'LIST_IP_POOLS',
    meta: {
      method: 'GET',
      url: '/ip-pools',
      showErrorAlert: false
    }
  });
}

export function createPool({ name }) {
  return sparkpostApiRequest({
    type: 'CREATE_IP_POOL',
    meta: {
      method: 'POST',
      url: '/ip-pools',
      data: { name }
    }
  });
}

export function updatePool(id, update) {
  return sparkpostApiRequest({
    type: 'UPDATE_IP_POOL',
    meta: {
      method: 'PUT',
      url: `/ip-pools/${id}`,
      data: update
    }
  });
}

export function deletePool(id) {
  return sparkpostApiRequest({
    type: 'DELETE_IP_POOL',
    meta: {
      method: 'DELETE',
      url: `/ip-pools/${id}`
    }
  });
}

export function getPool(id) {
  return sparkpostApiRequest({
    type: 'GET_IP_POOL',
    meta: {
      method: 'GET',
      url: `/ip-pools/${id}`
    }
  });
}
