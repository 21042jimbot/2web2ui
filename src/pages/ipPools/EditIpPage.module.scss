@import '~@sparkpost/matchbox/src/styles/config.scss';

.LineChartSection {
  padding: spacing('larger');
}
