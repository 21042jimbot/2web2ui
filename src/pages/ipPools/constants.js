export const IP_WARMUP_STAGES = [
  { name: 'Stage 1', id: 1, volume: 200 },
  { name: 'Stage 2', id: 2, volume: 500 },
  { name: 'Stage 3', id: 3, volume: 1000 },
  { name: 'Stage 4', id: 4, volume: 2000 },
  { name: 'Stage 5', id: 5, volume: 5000 },
  { name: 'Stage 6', id: 6, volume: 10000 },
  { name: 'Stage 7', id: 7, volume: 20000 },
  { name: 'Stage 8', id: 8, volume: 40000 },
  { name: 'Stage 9', id: 9, volume: 100000 },
  { name: 'Stage 10', id: 10, volume: 250000 },
  { name: 'Stage 11', id: 11, volume: 500000 },
  { name: 'Stage 12', id: 12, volume: 1000000 },
  { name: 'Stage 13', id: 13, volume: 2000000 },
  { name: 'Stage 14', id: 14, volume: 5000000 },
  { name: 'Stage 15', id: 15, volume: 10000000 },
  { name: 'Stage 16', id: 16, volume: 20000000 },
  { name: 'Stage 17', id: 17, volume: 40000000 },
  { name: 'Stage 18', id: 18, volume: 80000000 },
  { name: 'Stage 19', id: 19, volume: 160000000 },
  { name: 'Stage 20', id: 20, volume: 320000000 }
];
