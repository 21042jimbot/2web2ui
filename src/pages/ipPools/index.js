import IpPoolsList from './ListPage';

export default {
  ListPage: IpPoolsList
};
