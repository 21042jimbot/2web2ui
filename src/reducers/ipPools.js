const initialState = { list: [], listError: null };

export default (state = initialState, { type, payload }) => {
  switch (type) {

    //list pools
    case 'LIST_IP_POOLS_PENDING':
      return { ...state, listLoading: true, listError: null };

    case 'LIST_IP_POOLS_SUCCESS':
      return { ...state, list: payload, listLoading: false };

    case 'LIST_IP_POOLS_FAIL':
      return { ...state, listError: payload, listLoading: false };

    //create pool
    case 'CREATE_IP_POOL_PENDING':
      return { ...state, currentPool: { }, createError: null };

    case 'CREATE_IP_POOL_SUCCESS':
      return { ...state, currentPool: { id: payload.id, name: payload.name }};

    case 'CREATE_IP_POOL_FAIL':
      return { ...state, currentPool: {}, createError: payload };

    //update pool
    case 'UPDATE_IP_POOL_PENDING':
      return { ...state, updateLoading: true, updateError: null, updateSuccess: null };

    case 'UPDATE_IP_POOL_SUCCESS':
      return { ...state, updateSuccess: true, updateLoading: false };

    case 'UPDATE_IP_POOL_FAIL':
      return { ...state, updateLoading: false, updateError: payload };

    //get single pool
    case 'GET_IP_POOL_PENDING':
      return { ...state, getLoading: true };

    case 'GET_IP_POOL_SUCCESS':
      return { ...state, pool: payload, getLoading: false };

    case 'GET_IP_POOL_FAIL':
      return { ...state, pool: {}, getLoading: false, getError: payload };

    //delete a pool
    case 'DELETE_IP_POOL_PENDING':
      return { ...state, deleteLoading: true };

    case 'DELETE_IP_POOL_SUCCESS':
      return { ...state, deleteSuccess: true, deleteLoading: false };

    case 'DELETE_IP_POOL_FAIL':
      return { ...state, deleteLoading: false, deleteError: payload };

    default:
      return state;
  }
};
