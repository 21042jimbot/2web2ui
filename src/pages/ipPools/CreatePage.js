import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { Page } from '@sparkpost/matchbox';
import { ApiErrorBanner, Loading } from 'src/components';
import PoolForm from './components/PoolForm';

import { showAlert } from 'src/actions/globalAlert';
import { createPool, listPools } from 'src/actions/ipPools';


const breadcrumbAction = {
  content: 'IP Pools',
  Component: Link,
  to: '/account/ip-pools'
};

export class CreatePage extends Component {
  loadDependentData = () => {
    this.props.listPools();
  };

  componentDidMount() {
    this.loadDependentData();
  }

  createPool = (values) => {
    const { createPool, showAlert, history } = this.props;

    return createPool(values).then(() => {
      showAlert({
        type: 'success',
        message: `Created IP pool ${values.name}.`
      });
      history.push('/account/ip-pools');
    });
  };

  renderError() {
    const { listError } = this.props;
    return <ApiErrorBanner
      errorDetails={listError.message}
      message="Sorry, we seem to have had some trouble loading your IP pool."
      reload={this.loadDependentData}
    />;
  }

  render() {
    const { loading, listError } = this.props;

    if (loading) {
      return <Loading />;
    }

    if (listError) {
      return this.renderError();
    }

    return (
      <Page title="Create IP Pool" breadcrumbAction={breadcrumbAction}>
        <PoolForm onSubmit={this.createPool} isNew={true} />
      </Page>
    );
  }
}

const mapStateToProps = (state, props) => ({
  listError: state.ipPools.listError,
  loading: state.ipPools.listLoading
});

export default connect(mapStateToProps, {
  createPool,
  listPools,
  showAlert
})(CreatePage);
