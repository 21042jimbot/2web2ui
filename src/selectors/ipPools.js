import { createSelector } from 'reselect';
import _ from 'lodash';
import { currentPlanCodeSelector } from 'src/selectors/accountBillingInfo';
import { ENTERPRISE_PLAN_CODES } from 'src/constants';
import { isAdmin } from 'src/helpers/conditions/user';
const DEFAULT_POOL_ID = 'default';

const currentIpPoolId = (state, props) => props.match.params.poolId;
const currentSendingIp = (state, props) => props.match.params.ip;

export const getIpPools = (state) => state.ipPools.list;

export const selectCurrentPool = createSelector(
  [getIpPools, currentIpPoolId], (allPools, poolId) => _.find(allPools, { id: poolId }) || {}
);

export const selectIpsForCurrentPool = createSelector(
  [selectCurrentPool],
  ({ ips = []}) => ips
);

export const selectIpForCurrentPool = createSelector(
  [selectCurrentPool, currentSendingIp],
  ({ ips = []}, sendingIp) => _.find(ips, { external_ip: sendingIp })
);

export const getDefaultPool = createSelector(
  [getIpPools],
  (ipPools) => ipPools.find(({ id }) => id === DEFAULT_POOL_ID)
);

export const getNonDefaultIpPools = createSelector(
  [getIpPools],
  (ipPools) => ipPools.filter(({ id }) => id !== DEFAULT_POOL_ID)
);

export const getOrderedIpPools = createSelector(
  [getIpPools, getDefaultPool, getNonDefaultIpPools],
  (initialList, defaultPool, others) => {
    if (!defaultPool) {
      return initialList;
    }
    return [defaultPool, ...others];
  }
);

/**
 * Returns whether Purchase IP cta should shown, based on account's plan code
 * @return bool
 */
export const shouldShowIpPurchaseCTA = createSelector(
  [currentPlanCodeSelector, isAdmin], (currentPlanCode, admin) => !_.includes(ENTERPRISE_PLAN_CODES, currentPlanCode) && admin
);

export const selectFirstIpPoolId = createSelector(
  [getIpPools], (ipPools) => _.get(ipPools, '[0].id')
);

export const selectIpFormInitialValues = createSelector(
  [selectIpForCurrentPool, selectCurrentPool], (currentIp, pool) => ({
    auto_warmup_stage: 1,
    ...currentIp,
    ip_pool: pool.id

  })
);

/**
 * Returns whether overflow pool field is editable or not. If the current pool is an overflow pool for any other pool, overflow pool for current pool is not editable
 * @return bool
 */
export const canEditOverflowPool = createSelector(
  [getIpPools, selectCurrentPool], (pools, currentPool) => _.every(pools, (pool) => pool.auto_warmup_overflow_pool !== currentPool.id));
