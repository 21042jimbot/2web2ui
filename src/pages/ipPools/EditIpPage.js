import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page } from '@sparkpost/matchbox';
import _ from 'lodash';
import { ApiErrorBanner, Loading } from 'src/components';
import IpForm from './components/IpForm';
import { showAlert } from 'src/actions/globalAlert';
import { listPools, updatePool } from 'src/actions/ipPools';
import { updateSendingIp } from 'src/actions/sendingIps';
import { selectCurrentPool, selectIpForCurrentPool } from 'src/selectors/ipPools';

export class EditIpPage extends Component {
  onUpdateIp = values => {
    const { updateSendingIp, ip, showAlert, history } = this.props;

    if (!values.auto_warmup_enabled) {
      delete values.auto_warmup_stage;
    }

    const reAssignedPool = ip.ip_pool !== values.ip_pool;

    return updateSendingIp(ip.external_ip, values)
      .then(() => {
        if (reAssignedPool) {
          history.replace(`/account/ip-pools/edit/${values.ip_pool}/${ip.external_ip}`);
        }
        showAlert({
          type: 'success',
          message: `Updated IP ${ip.external_ip}.`,
        });
      })
      .then(this.loadDependentData);
  };

  loadDependentData = () => {
    this.props.listPools();
  };

  componentDidMount() {
    this.loadDependentData();
  }

  render() {
    const { loading, pool, ip, error } = this.props;

    if (loading || _.isEmpty(pool) || _.isEmpty(ip)) {
      return <Loading />;
    }

    return (
      <Page
        title={`Sending IP: ${ip.external_ip}`}
        breadcrumbAction={{
          content: pool.name,
          Component: Link,
          to: `/account/ip-pools/edit/${pool.id}`,
        }}
      >
        {error ? (
          <ApiErrorBanner
            errorDetails={error.message}
            message="Sorry, we seem to have had some trouble loading your IP data."
            reload={this.loadDependentData}
          />
        ) : (
          <IpForm onSubmit={this.onUpdateIp} />
        )}
      </Page>
    );
  }
}

const mapStateToProps = (state, props) => {
  const { listLoading, listError } = state.ipPools;

  return {
    ip: selectIpForCurrentPool(state, props),
    pool: selectCurrentPool(state, props),
    loading: listLoading,
    error: listError,
  };
};

export default connect(mapStateToProps, {
  updatePool,
  listPools,
  updateSendingIp,
  showAlert,
})(EditIpPage);
