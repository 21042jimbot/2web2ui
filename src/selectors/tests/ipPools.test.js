import { getNonDefaultIpPools, getDefaultPool, getOrderedIpPools } from '../ipPools';

describe('Selector: ipPools', () => {

  let state;

  beforeEach(() => {
    state = {
      ipPools: {
        list: [
          { id: 'none', ips: []},
          { id: 'big', ips: [1, 2, 3, 4, 5, 6]},
          { id: 'default', ips: [1, 2, 3]},
          { id: 'small', ips: [1, 2]}
        ]
      }
    };
  });

  it('should return the default pool', () => {
    expect(getDefaultPool(state)).toEqual({ id: 'default', ips: [1, 2, 3]});
  });

  it('should return a list of non default IP pools', () => {
    expect(getNonDefaultIpPools(state)).toEqual([
      { id: 'none', ips: []},
      { id: 'big', ips: [1, 2, 3, 4, 5, 6]},
      { id: 'small', ips: [1, 2]}
    ]);
  });

  it('should return a list of ordered pools (default first, then ordered by # of ips descending)', () => {
    expect(getOrderedIpPools(state).map((pool) => pool.id)).toEqual([
      'default', 'big', 'small', 'none'
    ]);
  });

});

