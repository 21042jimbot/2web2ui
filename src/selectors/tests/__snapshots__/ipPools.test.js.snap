// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector: ipPools selectReAssignPoolsOptions returns formatted pools list and renames label of current selected pool 1`] = `
Array [
  Object {
    "label": "-- Select a new pool --",
    "value": "big",
  },
  Object {
    "label": "None (none)",
    "value": "none",
  },
  Object {
    "label": "Default (default)",
    "value": "default",
  },
  Object {
    "label": "Small Pool (small)",
    "value": "small",
  },
]
`;
